src/zkevm_specs/evm/table.py

```diff
@@ -441,29 +461,38 @@ def __init__(
         )
         if copy_circuit is not None:
             self.copy_table = self._convert_copy_circuit_to_table(copy_circuit)
+        if keccak_table is not None:
+            self.keccak_table = set(keccak_table)
 
     def _convert_copy_circuit_to_table(self, copy_circuit: Sequence[CopyCircuitRow]):
-        rows = []
+        rows: List[CopyTableRow] = []
         for i, row in enumerate(copy_circuit):
-            if row.is_first != 1:
-                continue
-            assert i + 1 < len(copy_circuit), "Not enough rows in copy circuit"
-            next_row = copy_circuit[i + 1]
-            assert next_row.q_step == 0, "Invalid copy circuit"
-            rows.append(
-                CopyTableRow(
-                    src_id=row.id,
-                    src_type=row.tag,
-                    dst_id=next_row.id,
-                    dst_type=next_row.tag,
-                    src_addr=row.addr,
-                    src_addr_end=row.src_addr_end,
-                    dst_addr=next_row.addr,
-                    length=row.bytes_left,
-                    rw_counter=row.rw_counter,
-                    rwc_inc=row.rwc_inc_left,
+            # the first row and the row next to it will be used for its fields.
+            if row.is_first == 1:
+                first_row = row
+                assert i + 1 < len(copy_circuit), "Not enough rows in copy circuit"
+                next_row = copy_circuit[i + 1]
+                assert next_row.q_step == 0, "Invalid copy circuit"
+
+            # update `value_rlc` when we encounter the copy event's last row.
+            if row.is_last == 1:
+                value_rlc = row.value
+                rows.append(
+                    CopyTableRow(
+                        src_id=first_row.id,
+                        src_type=first_row.tag,
+                        dst_id=next_row.id,
+                        dst_type=next_row.tag,
+                        src_addr=first_row.addr,
+                        src_addr_end=first_row.src_addr_end,
+                        dst_addr=next_row.addr,
+                        length=first_row.bytes_left,
+                        value_rlc=value_rlc,
```

Comment:
Yes. This part was a bit tricky since the RLC accumulated value is only in the last row, but all other fields are available in the first row.

So for `row.is_first` I store `first_row` and `next_row` (which is the second row). And wait until we encounter `row.is_last`, where I get the `row.value` and populate the rest of the fields from `first_row` and `next_row`. What do you think about this approach? And in your opinion, is it possible to avoid this?

In the circuits repository, I can add the RLC accumulated value to [`CopyEvent`](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/execution.rs#L206) itself, so we won't need to do this dynamic rotation. Wdyt?

Reply:
> In the circuits repository, I can add the RLC accumulated value to [`CopyEvent`](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/execution.rs#L206) itself, so we won't need to do this dynamic rotation. Wdyt?

I think that still requires a dynamic rotation.

This is the current lookup expression:
```rust
impl<F: Field> LookupTable<F> for CopyCircuit<F> {
    fn table_exprs(&self, meta: &mut VirtualCells<F>) -> Vec<Expression<F>> {
        vec![
            meta.query_advice(self.is_first, Rotation::cur()),
            meta.query_advice(self.id, Rotation::cur()), // src_id
            self.tag.value(Rotation::cur())(meta),       // src_tag
            meta.query_advice(self.id, Rotation::next()), // dst_id
            self.tag.value(Rotation::next())(meta),      // dst_tag
            meta.query_advice(self.addr, Rotation::cur()), // src_addr
            meta.query_advice(self.src_addr_end, Rotation::cur()), // src_addr_end
            meta.query_advice(self.addr, Rotation::next()), // dst_addr
            meta.query_advice(self.bytes_left, Rotation::cur()), // length
            meta.query_advice(self.rw_counter, Rotation::cur()), // rw_counter
            meta.query_advice(self.rwc_inc_left, Rotation::cur()), // rwc_inc_left
        ]
    }
}
```
Which query would you do to get the `value_rlc`?  It would be `meta.query_advice(self.value, Rotation(n))` but `n` is not a fixed known value.

Reply:
Okay, I understand the problem now. In that case may be I can add an additional field `acc_rlc` which is `0` for all other copy data types except for `RlcAcc`. And in this case it will be the same over all rows, so we get an additional constraint of checking whether the `value == acc_rlc` for the last row.

Once this is in place, we only need to lookup the first/second row (which is 0 or 1 rotation) even for the `acc_rlc`

Reply:
> Okay, I understand the problem now. In that case may be I can add an additional field `acc_rlc` which is `0` for all other copy data types except for `RlcAcc`. And in this case it will be the same over all rows, so we get an additional constraint of checking whether the `value == acc_rlc` for the last row.
> 
> Once this is in place, we only need to lookup the first/second row (which is 0 or 1 rotation) even for the `acc_rlc`

This solution sounds correct!

Another idea that came to my mind which doesn't need a new column but complicates things is:
Instead of calculating the `acc_rlc` from top to bottom, do it from bottom to top.  So `acc_rlc[i] = acc_rlc[i+1] * r + v`, whichi leaves the `RLC(v)` in `acc_rlc[0]`.  The complication with this is that then we no longer have `RLC(reversed(input))` but `RLC(input)`.  And currently the bytecode circuit uses `RLC(reversed(input))`.  We could get back to this idea in the future as an optimization maybe.

So I think adding this new column which contains a copy of the last write value in all the steps in an event is a good way to go!

Reply:
I just pushed here: https://github.com/privacy-scaling-explorations/zkevm-specs/pull/235/commits/68ead11f2f40c706f3ecf11db21a02b3b6fbf591

Would be nice to get a review on this one :) @ed255

Reply:
I think I personally would prefer keeping the practice of `RLC(reversed(input))` uniform throughout the codebase, if it is not too much of an issue adding an additional column. This column is `0` over all copy data types, except only for the `RlcAcc` type. So additionally there are 2 constraints:
1. `acc_rlc` is the same over all rows
2. `acc_rlc == value` for the last row (in case `dst_type` is `RlcAcc`.